### DISCUSSION TITLE
Pytest problems

### DISCUSSION THREAD

**Discussion by saifmode** [DavidVujic/python-polylith#313] [Q&A] [answered]
Seasons greetings, all!

I'm having some import issues with pytest.

In my bases and component files, I use the truncated style of import, e.g. `from my_app.foo import my_func`, rather than `from bases.my_app.foo import my_func` or `from components.my_app.bar import whatever`.

This works well and my IDE can pick it up.

However, pytest is having issues with it. I tried setting my pytest options in `pyproject.toml` to:

```
[tool.pytest.ini_options]
pythonpath = ["bases", "components"]
addopts = ["--import-mode=importlib"]
```

But the pytest runner can only ever handle truncated imports from the first element in the pythonpath list.

My work around is to just do `pythonpath = "components"`, and then if a test needs to import something from a base file, I do `from bases.my_app.foo` instead of `from my_app.foo`.

But I wondered if anyone else has run into this issue, and knows a way of setting up the config so that I can use truncated style imports everywhere in my repo?

**Comment by DavidVujic** [comment]
Hello @saifmode! I think it looks correct with your way of importing without the `bases` or `components` prefix. Those shouldn't be needed at all, they are just like when using a `src` folder (but here we have two of them).

The way your IDE and the tool you use (`uv` if I remember it correctly) will be aware of those folders from the top-level `pyproject.toml` configuration (the _dev-mode-dirs_ part).

I don't think it is necessary to define the paths in a `pytest` config either, but maybe that is a use case I haven't run into yet.

By running pytest within the `uv` context, it should find all paths needed:
```shell
uv run pytest

# or if you use Poetry
poetry run pytest
```

Make sure the top-level `pyproject.toml` has this config (assuming that it is `uv` you are using). This is only for the top-level config and should not be in the project-specific toml files:
```shell
[tool.hatch.build]
dev-mode-dirs = ["components", "bases", "development", "."]
```

After adding this config, I think that you need to run `uv sync` once.

**Reply by saifmode** [reply, marked as answer]
Thanks for getting back! Unfortunately the above didn't fix the issue, as this is basically what I've already been doing.

However, I realised what the problem was. I had `__init__.py` files in the `bases/my_app` and `components/my_app` directories.

Deleting these resolved the issue.

**Reply by DavidVujic** [reply]
Good find! 👏 👏 👏  That would explain it. The `__init__.py` can cause problems in a Python namespace package structure.
